data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsDynamicVisibilityTimeoutManager.java

```diff
@@ -0,0 +1,123 @@
+
+        final ScheduledFuture<?> scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(
+                () -> increaseVisibilityTimeout(visibilityTimeoutIncrease, message),
+                initialWaitBeforeIncrease,
+                waitBeforeIncreasingAgain, TimeUnit.SECONDS);
+
+        messageToScheduledFuture.put(message.messageId(), scheduledFuture);
+    }
+
+    public void stopDynamicVisibilityForMessage(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            boolean cancelled = scheduledFutureToCancel.cancel(true);
+
+            if (!cancelled) {
+                LOG.error("Unable to cancel dynamic visibility execution for message {}", messageId);
+            } else {
+                messageToScheduledFuture.remove(messageId);
+            }
+        }
+    }
+
+    public int getRemainingTimeBeforeNextChangeMessageIncrease(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            return (int) scheduledFutureToCancel.getDelay(TimeUnit.SECONDS);
+        }
+
+        return initialMessageVisibilityTimeout;
+    }
+
+    public void shutdownDynamicMessaging() {
+        try {
+            scheduledExecutorService.shutdown();
+            scheduledExecutorService.awaitTermination(3, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            LOG.error("Dynamic SQS messaging was interrupted during shutdown");
+        } finally {
+            if (!scheduledExecutorService.isTerminated()) {
+                scheduledExecutorService.shutdown();
+            }
+        }
+    }
+
+    private void increaseVisibilityTimeout(final int visibilityTimeoutIncrease,
+                                           final Message message) {
+
+        final ChangeMessageVisibilityRequest changeMessageVisibilityRequest = ChangeMessageVisibilityRequest.builder()
+                .receiptHandle(message.receiptHandle())
+                .visibilityTimeout(visibilityTimeoutIncrease)
+                .queueUrl(queueUrl)
+                .build();
+
+        try {
+            sqsClient.changeMessageVisibility(changeMessageVisibilityRequest);
+        } catch (final Exception e) {
+            LOG.error("An exception occurred while changing the visibility timeout for message {}", message.messageId(), e);
+        }
+    }
+
+    private int calculateVisibilityTimeoutFromObjectSize(final long objectSizeBytes) {
+        // simple implementation to allocate 30 seconds per 0.25 GB of object, which means a minimum of 30 seconds added
+        final int visibilityTimeoutSeconds = (int) Math.ceil((double) objectSizeBytes / QUARTER_GB_BYTES) * VISIBILITY_TIMEOUT_SECONDS_PER_QUARTER_GB;
```

Comment:
Not sure if it's possible from an S3 perspective, but there's a potential edge case here with 0 byte objects